File: src/linked-lists/copy-list-with-random-pointer-138.py
class Solution(object):
    """
    """
    def copyRandomList(self, head):
        dic, prev, node = {}, None, head
        while node:
            if node not in dic:
                # Use a dictionary to map the original node to its copy
                dic[node] = Node(node.val, node.next, node.random)
            if prev:
                # Make the previous node point to the copy instead of the original.
                prev.next = dic[node]
            else:
                # If there is no prev, then we are at the head. Store it to return later.
                head = dic[node]
            if node.random:
                if node.random not in dic:
                    # If node.random points to a node that we have not yet encountered, store it in the dictionary.
                    dic[node.random] = Node(node.random.val, node.random.next, node.random.random)
                # Make the copy's random property point to the copy instead of the original.
                dic[node].random = dic[node.random]
            # Store prev and advance to the next node.
            prev, node = dic[node], node.next
        return head

File: src/linked-lists/copy-list-with-random-pointer-138.java

class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}

class Solution {
    public Node copyRandomList(Node head) {
        if(head == null)
            return null;
        Node cur = head, nextNode = null;
        
        //list becomes: origin - copy - origin - copy - origin - copy
        while(cur != null){
            nextNode = cur.next;
            cur.next = new Node(cur.val);
            cur.next.next = nextNode;
            cur = nextNode;
        }
        //assign copyNode random pointer
        cur = head;
        while(cur != null){
            if(cur.random != null){
                cur.next.random = cur.random.next;
            }
            cur = cur.next.next;
        }
        
        cur = head;
        Node copyHead = cur.next;
        
        //restore node
        while(cur != null){
            nextNode = cur.next.next;
            Node copyNode = cur.next;
            cur.next = nextNode;
            if(nextNode != null)
                copyNode.next = nextNode.next;
            cur = nextNode;
        }
        
        return copyHead;
        
    }
}